### ISSUE TITLE
[docs] Urls changed => forum links would become invalid

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#39007]
Originally in https://github.com/pytorch/pytorch/issues/18095#issuecomment-633149307:

@vadimkantorov:
Works: https://pytorch.org/docs/stable/torch.html#torch.flip
Breaks: https://pytorch.org/docs/master/torch.html#torch.flip

@t-vi:
They're now separate pages: https://pytorch.org/docs/master/generated/torch.flip.html#torch.flip
...but it would break many links, including on the forums.

@ezyang:
Yes, this will be a release blocker for next release.

cc @ezyang @gchanan @zou3519 @mattip

**Comment by mattip** [comment]
Hmm. Are the pages hosted in a way that we can do redirects to replace `torch.html#torch.\(.*\)` with `generated/torch.\1.html#torch.\1` ?

If so we can write a script to 
- scrape the stable website and pull out all the links
- scrape the master website and putll out all the links
- find the links that are no longer present on master
- try to redirect to the new locations

**Comment by vadimkantorov** [comment]
One can also do a server or client redirects as fallback

**Comment by vadimkantorov** [comment]
One way could also be eliminate “generated” folder from the urls...

**Comment by mattip** [comment]
> One way could also be eliminate “generated” folder from the urls...

Then we would still have to convert `torch.html#torch.flip` to `torch.flip.html#torch.flip`, note the extra `flip.`

**Comment by t-vi** [comment]
A not quite as neat but reasonable alternative could be to see to anchors being added to the (just the link) entries of the page currently served at that URL.

**Comment by ShawnZhong** [comment]
If I recall correctly, pytorch.org is served by GitHub pages, which does not allow server configurations (say URL rewrite), but client-side redirects should be an option. 

For adding anchors, I didn't see native support in Sphinx to generate anchors for `autosummary`. I guess we either need to hack into Sphinx or manually add them.

**Comment by vadimkantorov** [comment]
@ShawnZhong 
> Then we would still have to convert `torch.html#torch.flip` to `torch.flip.html#torch.flip`, note the extra `flip.`

Even in that case, maybe better to remove "generated"? Since it will be in every URL, it's a bit cleaner without it, and it carries no special meaning, docs are (almost) always generated

**Comment by ShawnZhong** [comment]
Not sure exactly why, but it seems that some popular python libs have `generated` in the docs: 
- https://numpy.org/doc/1.18/reference/generated/numpy.add.html
- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.randint.html
- http://ipython.org/ipython-doc/stable/api/generated/IPython.config.application.html

Maybe @mattip knows more about this.

**Comment by ShawnZhong** [comment]
One of the minor benefits I can think of is it's easier to add a single `generated` folder to `.gitignore` and to delete all generated files.

**Comment by vadimkantorov** [comment]
Well, I was just worried that in future if scheme changes, there would be a lot of hyperlinks on the internet containing that "generated" thing. But it's not much important

**Comment by vadimkantorov** [comment]
Also there would be a challenge to remove old pages from Google search which is quite nasty (https://github.com/pytorch/pytorch/issues/3428). It would be unfortunate, if Google search finds both new and old links. By the way, NumPy/SciPy do a bad job here...

**Comment by mattip** [comment]
> One of the minor benefits I can think of is it's easier to add a single generated folder to .gitignore and to delete all generated files

This seems to be the motivation, and the use of `generated` is standard across projects that use sphinx.

> A not quite as neat but reasonable alternative could be to add to anchors being added to the (just the link) entries of the page currently served at that URL.

If we post-process all the html, we could add an anchor for these entries:
- add an id for the tag
- add an anchor element to create the anchor link to the id.

For example `<p><a class="reference internal" href="generated/torch.flip.html#torch.flip"` ... would become `<p id=torch.flip><a class="headerlink" href="#torch.flip" title="Permalink">¶</a><a class="reference internal" href="generated/torch.flip.html#torch.flip"` ...

**Comment by mattip** [comment]
It is enough just to add the `id`, so 
```
<p><a class="reference internal" href="generated/torch.flip.html#torch.flip"
```
becomes
```
<p id="torch.flip"><a class="reference internal" href="generated/torch.flip.html#torch.flip"
```

**Comment by ShawnZhong** [comment]
As a side note here, if we just want to scroll to the method instead of redirecting to the detailed page, we can do: 

```js
document.querySelector('a[title="torch.flip"]').scrollIntoView()
```

**Comment by mattip** [comment]
I made sphinx emit references at the old location whenever `generated` is in the uri in gh-39086

**Comment by zou3519** [comment]
So we have two potential solutions:
1. Have javascript client-side redirect. There's a PR out here: #39032 
2. Add anchors to links on the page. There's a PR out here: #39086 

Honestly both seem reasonable to me (and equally as hacky). @mattip, @ShawnZhong, @vadimkantorov what do you folks think we should do?

**Comment by mattip** [comment]
Sorry, I didn’t see gh-39032. Either is fine.

**Comment by vadimkantorov** [comment]
@zou3519 I don't have preference, except that Google search results should not be spilled with both old and new URL versions

**Comment by mattip** [comment]
> Google search results should not be spilled with both old and new URL versions

Hmm. Not sure how to give a lower rank to the top-level documents than to the generated ones. It seems the tools available in robots.txt or meta tags are binary: either include or exclude whole pages. Perhaps if we use the javascript solution in gh-39032 but then exclude robots from seeing the javascript that would help? Of course users who turn off javascript would also not see the anchor-specific links and would be refered to the top of the pages ...

**Comment by vadimkantorov** [comment]
@zou3519 should have more insights about Google indexing hacking :)

**Comment by vadimkantorov** [comment]
@mattip about dichotomy between "top-level documents than to the generated ones.". I'm not sure I understand you. If by "top-level documents" you mean old pages, I think they should be phased out from Google search index, by first modifying their meta and simultaneously publishing new indexable pages. Like that, old links will continue to work and Google will not index old pages over time

**Comment by t-vi** [comment]
You'd want to have a list of available functions somewhere, ideally with link and one-line synopsis.

**Comment by vadimkantorov** [comment]
Well, maybe the "top-level" / index pages could be created manually, then old links could stay

**Comment by mattip** [comment]
The old links have become summaries. So the "old link" that points to this [whole documentation blurb](https://pytorch.org/docs/stable/torch.html#torch.allclose) on `stable`, now points to [the allclose line in the table](https://pytorch.org/docs/master/torch.html#comparison-ops) in `master` after this PR. Clicking on the `allclose` link brings you to the [full documentation](https://pytorch.org/docs/master/generated/torch.allclose.html#torch.allclose).  I wouldn;t want to totally disable indexing the summary page. since there is still useful information on it, for instance the documentation about [`torch.no_grad`](https://pytorch.org/docs/master/torch.html#locally-disabling-gradient-computation), so

**Comment by vadimkantorov** [comment]
The old link "https://pytorch.org/docs/master/torch.html#torch.flip" does not work (just checked).

I'm all for summary pages, but the new summary pages do not have right anchors, so the browser does not scroll down automatically, and this is problematic

**Comment by t-vi** [comment]
That's because #39086 isn't merged yet. :)

**Comment by vadimkantorov** [comment]
oh, makes sense :)

**Comment by vadimkantorov** [comment]
Some urls got broken again :( https://github.com/pytorch/pytorch.github.io/issues/599

**Comment by malfet** [comment]
@svekars do we have a mechanism to detect if new broken links are added to the docs?

**Comment by mikaylagawarecki** [comment]
fyi @svekars 

Summarizing discussion from triage meeting:
There are two types of broken links
1. broken links within our docs
2. external links that are broken

We should have a mechanism for detecting and eliminating (1). For (2), there are many answers in stack overflow that link to PyTorch docs, should we have a BC-surface on our docs links/redirect to ensure we don't break external links within some statute of limitations?

**Comment by svekars** [comment]
@malfet @mikaylagawarecki - we could look into enabling Sphinx linkchecker builder to identify broken cross-references. For external docs probably need to use something else - there is [sphinx-reredirects](https://documatt.com/sphinx-reredirects/index.html) extension for example. There should be a way to detect that a page is being moved in a PR and that a redirect must be created.

**Comment by vadimkantorov** [comment]
I think a problem is that for this anchor -> page factor-outs, the redirects should be done in a manual JavaScript hack comparing window.location.hash to some known list of redirects. Or at least, the anchors should be preserved so that https://pytorch.org/docs/stable/nn.html#torch.nn.Embedding scrolls to something reasonable (at least some table row)
